File: app/src/main/java/org/puffinbasic/runtime/PrintBuffer.java
package org.puffinbasic.runtime;

import it.unimi.dsi.fastutil.bytes.ByteArrayList;
import it.unimi.dsi.fastutil.bytes.ByteList;
import org.puffinbasic.file.PuffinBasicFile;

public class PrintBuffer {

    private static final byte SPACE = (byte) ' ';
    private final ByteList buffer;
    private int cursor;

    public PrintBuffer() {
        this.buffer = new ByteArrayList();
    }

    public void appendAtCursor(String value) {
        for (int i = buffer.size(); i < cursor + value.length(); i++) {
            buffer.add(SPACE);
        }
        for (int i = 0; i < value.length(); i++) {
            buffer.set(cursor++, (byte) value.charAt(i));
        }
    }

    public void flush(PuffinBasicFile file) {
        for (int i = 0; i < buffer.size(); i++) {
            file.writeByte(buffer.getByte(i));
        }
        for (int i = 0; i < buffer.size(); i++) {
            buffer.set(i, SPACE);
        }
        buffer.clear();
        cursor = 0;
    }
}

File: app/src/main/java/org/puffinbasic/runtime/PrintBuffer.kt
package org.puffinbasic.runtime

import org.puffinbasic.file.PuffinBasicFile

class PrintBuffer {
    private val buffer: MutableList<Byte> = mutableListOf()
    private var cursor = 0

    fun appendAtCursor(value: String) {
        for (i in buffer.size until cursor + value.length) {
            buffer.add(SPACE)
        }
        for (element in value) {
            buffer[cursor++] = element.code.toByte()
        }
    }

    fun flush(file: PuffinBasicFile) {
        for (i in buffer.indices) {
            file.writeByte(buffer[i])
        }
        for (i in buffer.indices) {
            buffer[i] = SPACE
        }
        buffer.clear()
        cursor = 0
    }

    companion object {
        private const val SPACE = ' '.code.toByte()
    }
}